bioio_ome_zarr/writers/ome_zarr_writer_v3.py

```diff
@@ -0,0 +1,541 @@
+import math
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import zarr
+from zarr.codecs import BloscCodec, BloscShuffle
+
+from .axes import Axes
+from .channel import Channel
+
+
+def spatial_downsample(
+    data: np.ndarray,
+    axes_names: List[str],
+    scale_factors: Tuple[int, ...],
+) -> np.ndarray:
+    """
+    Downsample the input array along spatial (X/Y) axes.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        Input array of dimensionality N. Only axes named "x" or "y"
+        are downsampled.
+    axes_names : List[str]
+        List of length N containing axis names (e.g.
+        ["t", "c", "z", "y", "x"]).
+    scale_factors : Tuple[int, ...]
+        Tuple of length N containing downsampling factors for each axis.
+
+    Returns
+    -------
+    np.ndarray
+        Array in which each "x" or "y" axis is downsampled by its
+        corresponding factor. Non-spatial axes remain unchanged.
+
+    Notes
+    -----
+    * Crop each spatial axis to a multiple of its factor before
+      reshaping.
+    * Compute the mean along each factor dimension to downsample.
+    """
+    names_lower = [n.lower() for n in axes_names]
+    shape = data.shape
+
+    crop_slices: List[Union[slice, int]] = []
+    reshape_dims: List[int] = []
+    factor_axes: List[int] = []
+    new_dim_idx = 0
+
+    for i, dim_size in enumerate(shape):
+        ax_name = names_lower[i]
+        f = scale_factors[i]
+
+        if ax_name in ("x", "y") and f > 1 and dim_size >= f:
+            full = (dim_size // f) * f
+            crop_slices.append(slice(0, full))
+            reshape_dims.extend([full // f, f])
+            factor_axes.append(new_dim_idx + 1)
+            new_dim_idx += 2
+        else:
+            crop_slices.append(slice(None))
+            reshape_dims.append(dim_size)
+            new_dim_idx += 1
+
+    data_cropped = data[tuple(crop_slices)]
+    data_reshaped = data_cropped.reshape(tuple(reshape_dims))
+    downsampled = data_reshaped.mean(axis=tuple(factor_axes))
```

Comment:
also, why not try to use the same downsampling system as the v2 reader?  does it not fit in with this writer?

Reply:
updated to use it in [ccc74a8](https://github.com/bioio-devs/bioio-ome-zarr/pull/41/commits/ccc74a8aaeb21603ebd6f2f4f34d7ce639df9037)